### DISCUSSION TITLE
Error with output printed out

### DISCUSSION THREAD

**Discussion by mhle510** [NASA-LIS/LISF#1424] [Q&A] [answered]
Hello,

I use the template provided below to print out the [LIS output](https://github.com/NASA-LIS/LISF/blob/master/lis/configs/MODEL_OUTPUT_LIST.TBL.adoc).

For the variable [SoilMoist], when I enable it to be printed out, I need to change its unit from kg/m2 to m3/m3.

There are other variables I want to print out, but they cause the program to stop. These are:

- GPP
- snowT
- RootMoist
- Vapor_Pressure

Could you please double-check these output variables?

**Comment by dmocko** [comment]
Hi @mhle510 

Can you please attach your MODEL OUTPUT TBL file and the lislog file from your stopped program?

**Reply by mhle510** [reply]
Morning David @dmocko 

Please find the attached file for specific MODEL OUTPUT and lislog related to the GPP error.
[output_log_forGPP.zip](https://github.com/NASA-LIS/LISF/files/12767752/output_log_forGPP.zip)

For the SnowT error, it is from the above Model Output, but with SnowT writing enabled. The lislog is similar to the GPP error, with differences only in the last lines
`[ERR] snowT is not defined 
[ERR] for diagnostic output...
[ERR] Please exclude it from the model output attributes table
 [ERR] Program stopping ..
 [ERR] endrun is being called
`

For other variables
`[ERR] RootMoist field is not defined
 [ERR] for diagnostic output...
 [ERR] Please exclude it from the model output attributes table
  [ERR] Program stopping ..
 [ERR] endrun is being called
`

`[ERR] Vapor_Pressure field is not defined
 [ERR] for diagnostic output...
 [ERR] Please exclude it from the model output attributes table
 [ERR] Program stopping ..
 [ERR] endrun is being called
`

**Comment by dmocko** [comment]
I moved this into a discussion, as it's not clear that there is an issue in the code.

What LSM are you using?  Note that some individual variables are not available as an output for every LSM.

I need to examine your files, and your lis.config file, to give a proper answer.

**Reply by mhle510** [reply]
I use Noah-MP.4.0.1. Please find attached the lis configuration for my run. Thanks

[lis.config_UMRB.txt](https://github.com/NASA-LIS/LISF/files/12767864/lis.config_UMRB.txt)

**Reply by mhle510** [reply]
One more note – If I don't print for all these four mentioned variables (GPP, snowT, RootMoist, Vapor_Pressure), my LIS run is fine.

**Reply by dmocko** [reply, marked as answer]
Hi @mhle510 

Sorry for the very slow response on this.  The reason that your GPP output is not working is that the MODEL OUTPUT TBL file is specifying units and direction that are not available as an output from the Noah-MP-4.0.1 LSM.  It's best to look into the code to see what options are available for each variable: 
https://github.com/NASA-LIS/LISF/blob/master/lis/surfacemodels/land/noahmp.4.0.1/NoahMP401_main.F90#L1119-L1121

Please try changing this line in your MODEL OUTPUT TBL file to:

`GPP:          1  g/m2s   IN   1 0 0 1 256 1       # Gross Primary Production`

I realize that our sample MODEL OUTPUT TBL for the GPP variable has the units/direction that you were trying.  I will investigate to see if these apply to other LSMs in LIS; otherwise, I will change the sample file.  Sorry for the trouble.

`SnowT` is not available to be output from Noah-MP-4.0.1 LSM.  However, `SnowTProf` is available, which will give you temperatures for all 3 snow layers.  Note that Noah-MP-4.0.1 has up to 3 snow layers (depending on total depth of the snow), and some layers may be undefined if there are only 2 (or 1, or 0) snow layers at that point at that time.  Turn off `SnowT` and change this line in your MODEL OUTPUT TBL to add `SnowTProf`:

`SnowTProf:    1  K       -    0 0 0 3 239 1000    # Temperature of the snow pack (K)`

https://github.com/NASA-LIS/LISF/blob/master/lis/surfacemodels/land/noahmp.4.0.1/NoahMP401_main.F90#L1022-L1026

`RootMoist` is not available as an output from Noah-MP-4.0.1 LSM, so you should leave it turned off.

Vapor Pressure (or `VaporPress_f`) would come from the forcing dataset, and not from the LSM (the `_f` at the end indicates it's from the forcing).  However, MERRA-2 and IMERG do not have this variable available to LIS.  Look for LIS_FORC_ variables in the "timeinterp" codes in these paths:
https://github.com/NASA-LIS/LISF/tree/master/lis/metforcing/merra2
https://github.com/NASA-LIS/LISF/tree/master/lis/metforcing/imerg

So, also turn off `VaporPress_f` in your MODEL OUTPUT TBL file.

Finally, if you want dynamic vegetation for Noah-MP-4.0.1, the LIS Users' Guide indicates that the Noah-MP developers generally recommend using option "5" instead of "2", so you could change this line in your lis.config file:

`Noah-MP.4.0.1 dynamic vegetation option:                                5  # Up to 10 different options`

**Reply by mhle510** [reply]
David, @dmocko 

No worries at all. Thank you so much for these details. I will follow your instructions.

Have a great day.

Hung
